Add git guide for mathlib4 contributors (#643)
leanprover-community/leanprover-community.github.io: templates/contribute/git.md

## Step 3: Set Up Remotes Correctly

The remote setup depends on which option you chose above:

### If you cloned your fork (Option A)

You need to add the original repository as `upstream`:

```bash
# Add the original repository as 'upstream'
git remote add upstream https://github.com/leanprover-community/mathlib4.git
```

### If you used an existing clone (Option B)

You need to rename the existing remote and add your fork:

```bash
# Rename the existing remote to 'upstream'
git remote rename origin upstream

# Add your fork as 'origin' (replace YOUR_USERNAME with your GitHub username)
git remote add origin https://github.com/YOUR_USERNAME/mathlib4.git
```

### Verify Your Remotes

Regardless of which option you chose, verify your remotes are set correctly:

```bash
git remote -v
```

You should see:
```
origin    https://github.com/YOUR_USERNAME/mathlib4.git (fetch)
origin    https://github.com/YOUR_USERNAME/mathlib4.git (push)
upstream  https://github.com/leanprover-community/mathlib4.git (fetch)
upstream  https://github.com/leanprover-community/mathlib4.git (push)
```

## Step 4: Configure the Master Branch

Make sure your `master` branch tracks `upstream/master`:

```bash

> git branch --set-upstream-to=upstream/master master

Reviewer: Does it work after `git fetch upstream`? 

This does look like a missing step in the guide, after doing `git remote add <remote-name>`, it is necessary to `git fetch <remote-name>`.

> git branch --set-upstream-to=upstream/master master

Reviewer 2: Yes, it works after `git fetch upstream`. Thanks!

> git branch --set-upstream-to=upstream/master master

Reviewer 3: Thanks for testing. I've adjusted the text accordingly.

```

## Step 5: Keep Your Master Branch Up to Date

Always keep your `master` branch synchronized with upstream:

```bash
git checkout master
git pull
```

Run these commands regularly, especially before creating new branches.

## Step 6: Prevent Accidental Commits to Master

To avoid accidentally committing directly to `master`, you can set up a pre-commit hook:

> ```bash

Reviewer 2: All these command-line code snippets are marked as `bash`. I guess most are one-liners that would work on any shell, but this one is clearly an exception. I'm unsure if it would work on `zsh`, which is the default shell on Mac OS, and it almost certainly wouldn't work on the Windows command line.

Author: Just verified this works fine on zsh.

I think the best we can do here is suggest googling for advice on how to set up a git hook on Windows. It is probably not easy unless you have a unix-like shell, and the advice is going to be different for every GUI that allows it (if any even do).

@loefflerd, could you investigate this and suggest a set of instructions that work for you?

Reviewer: Anyone who has installed Lean 4 following the official instructions has [Git for Windows](https://gitforwindows.org/) installed, which comes with Git Bash, which is, as the name suggests, a Unix-Like shell.

Reviewer 2: Looks like we need a  line`chmod +x .git/hooks/pre-commit` or something.

# Create the hook file
mkdir -p .git/hooks
cat > .git/hooks/pre-commit << 'EOF'
#!/bin/sh
branch="$(git rev-parse --abbrev-ref HEAD)"
if [ "$branch" = "master" ]; then
  echo "You can't commit directly to master branch"
  exit 1
fi
EOF

# Make it executable
chmod +x .git/hooks/pre-commit
```

## Step 7: Create and Configure New Branches

### Set Default Push Behavior

> Configure git to push new branches to `origin` by default:

Reviewer 3: Good point! I've reorganized the guide into two parts.

3. Click "Compare & pull request"
4. Fill in the PR title and description
5. Click "Create pull request"

Alternatively, if you have the GitHub CLI installed:

```bash
gh pr create --title "Your PR Title" --body "Description of your changes"
```

## Step 9: Working with Others' PRs

### Method 1: Manual Checkout

To check out someone else's PR manually:

```bash
# Add their fork as a remote (replace USERNAME with their GitHub username)
git remote add contributor-name https://github.com/USERNAME/mathlib4.git

# Fetch their branches
git fetch contributor-name

# Checkout their branch
git checkout contributor-name/their-branch-name
```

### Method 2: Using GitHub CLI (Recommended)

This is much simpler:

```bash
# Checkout PR #1234
gh pr checkout 1234
```

This automatically handles the remote setup and branch checkout.

## Step 10: Giving Collaborator Access

If you want to allow others to push directly to your PR branch:

1. Go to your fork: `https://github.com/YOUR_USERNAME/mathlib4`
2. Click "Settings" tab
4. Click "Collaborators" (you may have to reauthenticate)
5. Enter their GitHub username
6. Choose "Write" permission level
7. Send the invitation

> Once they accept, they can push directly to your PR branches.

Reviewer 3: ```suggestion
Once they accept, they can push directly to your PR branches by using `git push` after following one of the methods in step 9.
```

## ⚠️ Security Warning

**Important**: When you give someone collaborator access to your fork, or when you checkout and run someone else's code, you are potentially running unreviewed code on your computer. Only collaborate with people you trust, as they could potentially include malicious code that runs during the build process.

## Getting Help

If you encounter issues or have questions about git workflows, please ask in the `#new users` stream on the [Lean Zulip chat](https://leanprover.zulipchat.com). The community is very helpful and welcomes questions from newcomers!

## Quick Reference

Here's a summary of the most common commands you'll use:

```bash

> # Update master and push to your fork

Reviewer 3: ```suggestion
# Update master
```

```bash
git branch --set-upstream-to=upstream/master master
```

## Step 5: Configure Git for Your Workflow

### Set Default Push Behavior

Configure git to push new branches to `origin` by default:

```bash
git config push.default current
git config push.autoSetupRemote true
```

### Prevent Accidental Commits to Master (Optional but Recommended)

To avoid accidentally committing directly to `master`, you can set up a pre-commit hook. First, create the hook file:

```bash
mkdir -p .git/hooks
cat > .git/hooks/pre-commit << 'EOF'
#!/bin/sh
branch="$(git rev-parse --abbrev-ref HEAD)"
if [ "$branch" = "master" ]; then
  echo "You can't commit directly to master branch"
  exit 1
fi
EOF
```

Then make it executable:

```bash
chmod +x .git/hooks/pre-commit
```

---

# Part 2: Daily Workflow

These are the operations you'll use regularly when working on contributions.

## Creating and Working on a New Branch

### Keep Your Master Branch Up to Date

**Do this before creating a new branch** to ensure you're working with the latest changes:

```bash
git switch master
git pull
```

### Create a New Branch

Then create and switch to a new branch:

```bash
git switch -c my-feature-branch

> ```

Reviewer 3: Would you want to create a suggestion of what you have in mind?

> ```

Reviewer 4: Done below!

The branch will automatically track `origin/my-feature-branch` when you first push it.

## Push Your Branch and Open a PR

### Push Your Branch

After making your changes and commits:

```bash
git push origin my-feature-branch
```

### Open a Pull Request

1. Go to your fork on GitHub: `https://github.com/YOUR_USERNAME/mathlib4`
2. You should see a banner suggesting to open a PR for your recent push
3. Click "Compare & pull request"
4. Fill in the PR title and description
5. Click "Create pull request"

> 

Reviewer 4: ```suggestion

Alternatively, or if you do not see the banner, you can also go to https://github.com/leanprover-community/mathlib4/compare, and click `compare across fork`, you will need to select your fork in the "head repository" drop-down menu, and select the branch you want to merge in the "compare" drop-down menu.
```

> Alternatively, if you have the GitHub CLI installed:
> 
> ```bash
> gh pr create --title "Your PR Title" --body "Description of your changes"
> ```

Reviewer 3: ```suggestion
```

> 

Reviewer 3: ```suggestion
```

## Working with Others' PRs

### Method 1: Manual Checkout

To check out someone else's PR manually, first add their fork as a remote (replace `USERNAME` with their GitHub username):

```bash
git remote add contributor-name https://github.com/USERNAME/mathlib4.git
```

Then fetch their branches:

```bash
git fetch contributor-name
```

Finally, checkout their branch:

```bash
git checkout contributor-name/their-branch-name
```

### Method 2: Using GitHub CLI (Recommended)

This is much simpler. To checkout PR #1234:

```bash

> gh pr checkout 1234

Reviewer 3: Or rather, we should teach how to set the default. So that people can internalize the short command.

> gh pr checkout 1234

Reviewer 3: Great! Thanks for testing. Then I'll add the instruction.

> gh pr checkout 1234

Reviewer 3: Done

5. Enter their GitHub username
6. Choose "Write" permission level
7. Send the invitation

Once they accept, they can push directly to your PR branches by using `git push` after following one of the methods in step 9.

---

# Additional Information

## ⚠️ Security Warning

**Important**: When you give someone collaborator access to your fork, or when you checkout and run someone else's code, you are potentially running unreviewed code on your computer. Only collaborate with people you trust, as they could potentially include malicious code that runs during the build process.

## Getting Help

If you encounter issues or have questions about git workflows, please ask in the `#new users` stream on the [Lean Zulip chat](https://leanprover.zulipchat.com). The community is very helpful and welcomes questions from newcomers!

## Quick Reference

Here's a summary of the most common commands you'll use.

Update master:

```bash
git switch master
git pull
```

Create a new branch:

```bash
git switch -c new-branch-name
```

Push your branch and set up tracking:

```bash
git push origin new-branch-name
```

Check out someone else's PR:

```bash

> gh pr checkout PR_NUMBER

Reviewer 3: We now suggest setting the repo as default.

## Push Your Branch and Open a PR

### Push Your Branch

After making your changes and commits:

```bash
git push origin my-feature-branch
```

### Open a Pull Request

1. Go to your fork on GitHub: `https://github.com/YOUR_USERNAME/mathlib4`
2. You should see a banner suggesting to open a PR for your recent push
3. Click "Compare & pull request"
4. Fill in the PR title and description
5. Click "Create pull request"

## Working with Others' PRs

### Method 1: Manual Checkout

To check out someone else's PR manually, first add their fork as a remote (replace `USERNAME` with their GitHub username):

```bash
git remote add contributor-name https://github.com/USERNAME/mathlib4.git
```

Then fetch their branches:

```bash
git fetch contributor-name
```

Finally, checkout their branch:

```bash
git checkout contributor-name/their-branch-name
```

(Remotes can be removed with `git remote remove <contributor-name>`.)

### Method 2: Using GitHub CLI (Recommended)

This is much simpler. To checkout PR #1234:

```bash
gh pr checkout 1234
```

This automatically handles the remote setup and branch checkout.

> 

Reviewer 4: Only thing I slightly worry about now is that it being in the "Others' PR" section, some people might think this can’t apply to their own PRs as well...

> 

Reviewer 3: Should it move to the `Create a new branch` section?

> 

Reviewer 4: If it moves, then point `1` could be modified to 
```
1. Check out the relevant PR branch by running `git switch <pr-branch-name>` if this is your own PR, or by following the instructions in the section `Working with Others' PRs` if you intend to work on top of someone else's PR.
```

# Git Guide for Mathlib4 Contributors

> This guide is designed for mathematicians who are new to git but want to contribute to mathlib4. We'll walk through the essential git workflows step by step.

Reviewer 5: ```suggestion
This guide is designed for mathematicians who are new to git but want to contribute to the mathlib4 library.
Contributions are made via pull requests. We'll walk through the essential workflows step by step.
Note that there are many other guides on the web describing how to contribute to open-source projects
via pull requests.
```
I'm just wondering whether we should introduce the key phrase "pull request" early because this is what people can search for.

The guide is organized into two main sections:
1. **One-time setup** (do this once when you first start contributing)
2. **Daily workflow** (common operations for working on contributions)

## Prerequisites

Before starting, make sure you have:
- Git installed on your computer
- A GitHub account

> - (Optional but recommended) The GitHub CLI tool (`gh`) installed

Reviewer 6: ```suggestion
- (Optional but recommended) [The GitHub CLI tool (`gh`)](https://cli.github.com/) installed
```

---

# Part 1: One-Time Setup

These steps only need to be done once when you first start contributing to mathlib4.

## Step 1: Fork the Repository on GitHub

First, you need to create your own copy (fork) of the mathlib4 repository:

1. Go to https://github.com/leanprover-community/mathlib4
2. Click the "Fork" button in the top right corner
3. Choose your GitHub account as the destination. It is recommended to leave "copy the master branch only" checked.
4. Wait for GitHub to create your fork

**You only ever need to do this step once.** You can reuse your fork for many different branches and pull requests.

## Step 2: Get a Local Copy of the Repository

You have two options depending on whether you already have a clone of mathlib4:

### Option A: If you don't have mathlib4 cloned yet

Clone your fork (not the original repository). Replace `YOUR_USERNAME` with your GitHub username:

```bash

> git clone https://github.com/YOUR_USERNAME/mathlib4.git

Reviewer 5: I think the original is easier to understand for beginners.

git remote add upstream https://github.com/leanprover-community/mathlib4.git
```

### If you used an existing clone (Option B)

You need to rename the existing remote and add your fork. Replace `YOUR_USERNAME` with your GitHub username:

```bash
git remote rename origin upstream
git remote add origin https://github.com/YOUR_USERNAME/mathlib4.git
```

### Verify Your Remotes

Regardless of which option you chose, verify your remotes are set correctly:

```bash
git remote -v
```

You should see:
```
origin    https://github.com/YOUR_USERNAME/mathlib4.git (fetch)
origin    https://github.com/YOUR_USERNAME/mathlib4.git (push)
upstream  https://github.com/leanprover-community/mathlib4.git (fetch)
upstream  https://github.com/leanprover-community/mathlib4.git (push)
```

## Step 4: Configure the Master Branch

First, fetch the branches from upstream:

```bash
git fetch upstream
```

Then make sure your `master` branch tracks `upstream/master`:

```bash
git branch --set-upstream-to=upstream/master master
```

## Step 5: Configure Git for Your Workflow

### Set Default Push Behavior

Configure git to push new branches to `origin` by default:

```bash
git config push.default current

> git config push.autoSetupRemote true

Reviewer 6: They seem to be distinct options to me: https://git-scm.com/docs/git-push#Documentation/git-push.txt-pushautoSetupRemote

I'm not sure how much explanation we should give here; I'll leave this text alone for now.

```

### Configure GitHub CLI (if installed)

If you have the GitHub CLI installed, set the default repository to the upstream mathlib4:

```bash
gh repo set-default leanprover-community/mathlib4
```

This ensures that GitHub CLI commands like `gh pr checkout` will work with the main mathlib4 repository rather than your fork.

### Prevent Accidental Commits to Master (Optional but Recommended)

To avoid accidentally committing directly to `master`, you can set up a pre-commit hook. First, create the hook file:

```bash
mkdir -p .git/hooks
cat > .git/hooks/pre-commit << 'EOF'
#!/bin/sh
branch="$(git rev-parse --abbrev-ref HEAD)"
if [ "$branch" = "master" ]; then
  echo "You can't commit directly to master branch"
  exit 1
fi
EOF
```

Then make it executable:

```bash
chmod +x .git/hooks/pre-commit
```

---

# Part 2: Daily Workflow

These are the operations you'll use regularly when working on contributions.

## Creating and Working on a New Branch

### Keep Your Master Branch Up to Date

**Do this before creating a new branch** to ensure you're working with the latest changes:

```bash
git switch master

> git pull

Reviewer 5: I don't know much about git myself but I've been writing `git checkout master` for years. Is `switch` better?

```

### Create a New Branch

Then create and switch to a new branch:

```bash
git switch -c my-feature-branch
```

The branch will automatically track `origin/my-feature-branch` when you first push it.

## Push Your Branch and Open a PR

### Push Your Branch

After making your changes and commits:

```bash

> git push origin my-feature-branch

Reviewer 7: Is `origin` needed here?

```

### Open a Pull Request

1. Go to your fork on GitHub: `https://github.com/YOUR_USERNAME/mathlib4`
2. You should see a banner suggesting to open a PR for your recent push
3. Click "Compare & pull request"
4. Fill in the PR title and description
5. Click "Create pull request"

Alternatively, or if you do not see the banner, you can also go to https://github.com/leanprover-community/mathlib4/compare, and click `compare across forks`. You will need to select your fork in the "head repository" drop-down menu, and select the branch you want to merge in the "compare" drop-down menu.

> ## Working with Others' PRs

Reviewer 7: Here we should mention *very clearly* that even just opening VS Code will execute code written by a random person on the internet, with ZERO (*ZERO*) control by the mathlib community.

Reviewer 6: Ah, I see now that there's some text about security in the "Additional Information" section.

### Method 1: Manual Checkout

To check out someone else's PR manually, first add their fork as a remote (replace `USERNAME` with their GitHub username):

```bash
git remote add contributor-name https://github.com/USERNAME/mathlib4.git
```

Then fetch their branches:

```bash
git fetch contributor-name
```

Finally, checkout their branch:

```bash
git checkout contributor-name/their-branch-name
```

(Remotes can be removed with `git remote remove <contributor-name>`.)

> 

Reviewer 6: To switch back to your branch:
```bash
git switch -c my-feature-branch
```

### Method 2: Using GitHub CLI (Recommended)

This is much simpler. To checkout PR #1234:

```bash
gh pr checkout 1234
```

This automatically handles the remote setup and branch checkout.

## Giving Collaborator Access

If you want to allow others to push directly to your PR branch:

1. Go to your fork: `https://github.com/YOUR_USERNAME/mathlib4`
2. Click "Settings" tab
4. Click "Collaborators" (you may have to reauthenticate)
5. Enter their GitHub username
6. Choose "Write" permission level
7. Send the invitation

> Once they accept, they can push directly to your PR branches by using `git push` after following one of the methods in step 9.

Reviewer 6: ```suggestion
Once they accept, they can push directly to your PR branches by using `git push` after following one of the methods in ["Basing Work on an Other PR"](#basing-work-on-an-other-pr).
```
(Not sure if the link will work)

---

# Additional Information

## ⚠️ Security Warning

**Important**: When you give someone collaborator access to your fork, or when you checkout and run someone else's code, you are potentially running unreviewed code on your computer. Only collaborate with people you trust, as they could potentially include malicious code that runs during the build process.

## Getting Help

If you encounter issues or have questions about git workflows, please ask in the `#new users` stream on the [Lean Zulip chat](https://leanprover.zulipchat.com). The community is very helpful and welcomes questions!

## Quick Reference

Here's a summary of the most common commands you'll use.

Update master:

```bash
git switch master
git pull
```

Create a new branch:

```bash
git switch -c new-branch-name
```

Push your branch and set up tracking:

```bash
git push origin new-branch-name
```

Check out someone else's PR:

```bash
gh pr checkout PR_NUMBER
```

> 

Reviewer 6: ````suggestion
Switch back to working on your branch:

```bash
git switch your-branch-name
```
````
```
